paymentsheet/src/main/java/com/stripe/android/common/coroutines/CoalescingOrchestrator.kt

```diff
@@ -0,0 +1,80 @@
+package com.stripe.android.common.coroutines
+
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.CoroutineStart
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+
+/**
+ * This class intends to simplify workflows where calling a factory multiple times in parallel only results in a single
+ * call to the underlying factory, but the data returned is shared between all callers.
+ */
+internal class CoalescingOrchestrator<T : Any>(
+    private val factory: suspend () -> T,
+    private val keepDataInMemory: (T) -> Boolean = { false },
+    // This should only be used for testing.
+    private val awaitListener: (() -> Unit)? = null,
+) {
+    @Volatile
+    private lateinit var data: T
+
+    @Volatile
+    private var dataInitialized: Boolean = false
+
+    @Volatile
+    private var deferred: Deferred<T>? = null
+    private val lock: Any = Any()
+
+    tailrec suspend fun get(): T {
+        if (dataInitialized) {
+            return data
+        }
+
+        val result = coroutineScope {
+            val deferredToAwait: Deferred<T>
+            synchronized(lock) {
+                if (dataInitialized) {
+                    return@coroutineScope data
+                }
+                val localDeferred = deferred
+                if (localDeferred != null && !localDeferred.isCancelled) {
+                    deferredToAwait = localDeferred
+                } else {
```

Comment:
For my understanding, this can happen if the scope that owns the deferred call is canceled correct?

Reply:
I think I'm missing some context. Could you expand your question?

Reply:
Interesting. This isn't a scenario I thought about. Let me play with it.

Reply:
Yup, that's correct!